unsupported/asf_ops_R2.1.2/rds/src/rds/rds.man
.pl 66
.\"
.\"  @(#)rds.man	1.5 96/07/19 11:52:39
.\"
.TH RDS 1 "29 April 1996" "ASF SPS Release 1B'" "ASF USER COMMANDS"
.SH NAME
rds \- Raw Data Scanner (RDS) subsystem executable
.SH SYNOPSIS
.B rds
.B -asfn
.I name
.B -asf_CP_inet
.I CP-host-address
.B -asfp
.I TCP-port
.B -configfile
.I config-file
.B -config
.I msgTemplate-file
.B -asflogloc
.I location
.SH "DESCRIPTION"
.B rds
scans and decodes raw radar signal data as part of the SAR Processing
System (SPS).
.B Scanning
involves determination of the fixed-center image frames that are contained
within a given data acquisition and calculation of the locations of those
frames.  The output of a scanning job is a Scan Results file.
.B Decoding
involves unscrambling and reformatting raw radar signal, auxiliary,
replica, and ephemeris data into formats suitable for ingestion by the ScanSAR
Processor (SSP).
The output of a decoding job consists of five files: the Echo Data
file, Auxiliary Data file, Replica Data file, Ephemeris Data file, and the
Burst Offset Data file.  \fBrds\fP' external interface is via the Control
Processor (CP); jobs and commands are sent to \fBrds\fP over a socket
connection it establishes with \fBCP\fP.
.B rds
does not operate in standalone mode for the R1B' release.

.SS "Supported Formats"
.B rds
supports scanning of RADARSAT, ERS-1, and ERS-2 data.
For the R1B'
delivery, only decoding of RADARSAT ScanSAR data is supported, and JERS-1
scanning is not supported.

.B rds
supports raw input from disk files as well as from DCRSi tape drives
using the VDI controllers.  For R1B', \fBrds\fP only supports DCRSi firmware
version 4.4, although the software has been successfully used with version
5.0 drives.
.SH OPTIONS
.IP "\fB\-asfn\fP \fIname\fP"
Logical name, or label, to be associated with this instance of \fBrds\fP.
The name must begin with the characters \fIRDS\fR.
This item must be specified on the command line.
.IP "\fB\-asf_CP_inet\fP \fICP-host-address\fP"
IP address (dot-notation) of the machine on which \fBCP\fP is hosted.
This item must be specified on the command line.
.IP "\fB\-asfp\fP \fITCP-port\fP"
TCP port number used by \fICP\fR to listen for a connection from \fBrds\fP.
This item must be specified on the command line.
.IP "\fB\-configfile\fP \fIRDS-config-file\fP"
Name of the RDS configuration file (e.g. \fI$LOCAL/rds/rds-config.odl\fR)
that defines the configuration of this instance of \fBrds\fP.
This item must be specified on the command line.
.IP "\fB\-config\fP \fImsgTemplate-file\fP"
Name of the message template file that defines the formats of all ODL
messages recognized by \fBrds\fP
(i.e. \fI$LOCAL/sps/templates/msgTemplate.odl\fP).
This item must be specified on the command line.
.IP "\fB\-asflogloc\fP \fIlocation\fP"
Specifies the target location(s) for log messages.  Acceptable values
are \fISYSLOG\fR and \fISYSLOG_AND_TERM\fR.  If \fISYSLOG\fR is specified,
log messages will be written only to the syslog facility.  If
\fISYSLOG_AND_TERM\fR is specified, then the messages are also written
to the standard error.
This item must be specified on the command line.

.SH "CONFIGURATION PARAMETERS"
The \fBrds\fP configuration file contains the following tunable parameters
that may be adjusted under warranted conditions.  As these parameters
already carry quite generous defaults, their values should only be tightened
up and not further relaxed.
.IP "\fBSEEK_RETRY\fP"
Specifies the maximum number of times that RDS will execute the address-seek
sequence before declaring a seek error ("can't seek to block #N...").
Unexpected response or no response within \fBSEEK_TIMER\fP seconds from the
tape drive constitutes a retry failure.
.IP "\fBSEEK_TIMER\fP"
Specifies the maximum number of seconds that RDS will wait for a desired 
response from the tape drive before declaring a retry failure.
.IP "\fBTAPEBLK_MAX\fP"
Specifies the maximum number of tape blocks that \fBrds\fP scans or decodes
per job request.  The theoretical numbers for one 15-minute datatake and
one ScanSAR image frame are approximately \fI2650000\fR and \fI250000\fR,
respectively.  This parameter controls the amount of memory \fBrds\fP uses
at run-time. So, adjust it wisely. The recommended setting is \fI2000000\fR.
.IP "\fBTHREAD_MAX\fP"
Specifies the number of parallel scanning/decoding threads to be used.
For best performance, this parameter should be set to at least \fI2\fR,
which is the recommended setting.
.IP "\fBTAPEBUF_MAX\fP"
Specifies the number of I/O buffers to be shared and circulated among \fBrds\fP
threads.  As a minimum, there should be at least one buffer per thread.
However, \fBrds'\fP disk writer thread, which is activated only for decoding
jobs, requires at least two buffers.
\fBrds\fP uses one tape reader thread, one
disk writer thread, and THREAD_MAX scan/decode threads.  Thus, the minimum 
recommended setting is \fI6\fR.
.IP "\fBPPR_MAX\fP"
Specifies the maximum number of image frames that \fBrds\fP generates per 
job request, irrespective of the number of datatakes discovered.
The theoretical maximum number of image frames that could be generated
per datatake is \fI100\fR, which is the recommended setting.  This parameter
affects RDS total memory usage, so adjust it wisely.
.IP "\fBBURST_PER_IMAGE_MAX\fP"
Specifies the maximum number of bursts allowed per ScanSAR image frame.
The current implementation limit within the ScanSAR Processor is \fI1600\fR,
which is the recommended setting.  This parameter affects RDS total 
memory usage, so adjust it wisely.
.IP "\fBDWP_PER_BURST_MAX\fP"
Specifies the maximum number of data window position changes allowed per
burst.  The recommended setting is \fI4\fR.
.IP "\fBAGC_PER_BURST_MAX\fP"
Specifies the maximum number of AGC changes allowed per burst.
The recommended setting is \fI8\fR.
.IP "\fBDCRSI_CMD_CSIZE\fP"
Specifies the character size of the serial line (\fI/dev/vdi_sp0\fR)
used for DCRSi
command and response communication.  The allowable values are 7 or 8 bits.
A value of \fI7\fR should be used for the 4.4 drive, while \fI8\fR should
be used with a 5.0 drive.
.IP "\fBFRM_SYNC_MISS_MAX\fP"
Specifies the maximum number of corrupted bits allowed in a telemetry
frame SYNC code (32 bits for RADARSAT, and 24 bits for ERS-1/2), before a
SYNC miss is declared, and hence the end of a segment.  To begin a segment,
however, \fBrds\fP does require an exact SYNC code match.  It is assumed that
the value specified is for RADARSAT.  To arrive at the maximum corrupted bits
for for ERS-1/2 jobs, \fBrds\fP takes 3/4 of the specified value.  
The currently recommended setting is \fI24\fR bits, which permits up to 75%
corruption in the frame SYNC code.
.IP "\fBAUX_SYNC_MISS_MAX\fP"
Specifies the maximum number of corrupted bits allowed in a 32-bit Auxiliary
(AUX) SYNC code.  Detection of such AUX SYNC code signals the beginning of a
pulse.  This parameter applies only to RADARSAT jobs.  The recommended
setting is \fI3\fR bits.
.IP "\fBSCID_MISS_MAX\fP"
Specifies the maximum number of corrupted bits allowed in the 8-bit Spacecraft
Identifier field of an RADARSAT telemetry frame header.  The recommended
setting is \fI8\fR, which tells \fBrds\fP not to check this field.
.IP "\fBVALID_FLAG_MISS_MAX\fP"
Specifies the maximum number of corrupted bits allowed in a 11-bit Data Valid
Pointer field of an RADARSAT telemetry frame header.  The recommended
setting is \fI11\fR, which tells \fBrds\fP to disable the check of this field.
.IP "\fBSC_TIME_DELTA\fP"
Specifies the maximum difference in AUX spacecraft time in seconds between two
consecutive RADARSAT pulses before a gap is declared, hence terminating a
segment.  The recommended value is \fI3.8e-04\fR seconds.
.IP "\fBCAL_N_LINES\fP"
Specifies the number of range lines (pulses) to average in calculating average
caltone power in dB.  The allowable range of values is from \fI6\fR to \fI12\fR.
The recommended setting is \fI12\fR.
.IP "\fBCAL_DC_MASK\fP"
Specifies the percentage of spectrum to exclude when analyzing the caltone
data.  For instance, if CAL_DC_MASK = \fI0.10\fR, then 0.10*8192/2 bins at each
end of the caltone spectrum will be excluded from the average caltone power
calculation.  The allowable range is from \fI0.03\fR to \fI0.10\fR.
The recommended setting is \fI0.10\fR.
.IP "\fBCAL_THRESHOLD_RATIO\fP"
Specifies the cutoff ratio (s/m) of the standard deviation of the spectrum to
the mean, i.e., if s/m > CAL_THRESHOLD_RATIO, then \fBrds\fP
includes the range line
(pulse) in consideration in the average caltone power calculation; otherwise
\fBrds\fP excludes it.  The recommended value is \fI2.0\fP.
.\".SH "RETURN VALUE"

.SH MESSAGES
This section describes the error messages that are most likely encountered
during normal operation, beginning first with the most frequently occurred
errors observed during RDS development and testing.
Following each error message is a brief explanation of possible causes and 
recommended actions.
Debugging messages, which are quietly logged and which may be 
viewed using the
.B log_browser
utility, are also covered here.
.B rds
uses the \fIlocal4\fR syslog facility.
.IP "\fBCan't seek to block #N\fP"
There are several causes for this error. (1) The most likely is the
DCRSi tape drive simply can't go to this location because the tape medium 
itself is bad or there was no data recorded about this location.  To confirm
this, look at the RDS debug log for the DCRSi response to each play command 
\fI;PL AN;\fR, where \fIN\fR denotes the starting block address to seek.
If the 
\fIDL\fR string immediately before the last \fIDS\fR response contains a
non-zero address other than the specified address, 
then indeed this is the problem.
The recommended action in this case is to change the datatake's start address.
However, if the \fIDL\fR string is \fIDL B-00000000,00000000000\fR, then
the Error Correction Unit (ECC) is either disconnected from the DCRSi or not
powered on.  (2) Otherwise, \fBrds\fP' pre-programmed response wait time,
which is currently 90 seconds per command issued to the DCRSi,
may be too short.  
\fBrds\fP will issue the \fIPL\fR command twice and wait 90 seconds each
time for the desired response.  It is possible, though unlikely, 
that \fBrds\fP will time
out after 2 attempts without the tape drive reaching the desired location.
To confirm this scenario, look again at the last \fIDS\fR response
received from 
the DCRSi.  If it is not \fIDS C022\fR, \fIDS 4011\fR, or \fIDS 4000\fR,
then the tape drive is still trying to go to that location.
(3) Finally, the cause may be as silly as the DCRSi drive itself
is not powered-on or its cable connection is broken.  This problem
can be quicly identified by examining the DCRSi response to the \fIPL\fR
command.  If there is no response, then it is very likely that this is 
the cause of the error. (4) Otherwise, if the response appears to be 
undecipherable, random characters, then the problem is likely caused
by the mismatch betwen the DCRSi switch settings and the RDS
configuration parameter \fBDCRSI_CMD_CSIZE\fP.  The two must agree.
.IP "\fBScan results file created but no image frames generated\fP"
There are three possible causes for this error.  (1) The most obvious is 
the area scanned simply does not contain any data of desired mode. (2)
Another cause may be that the job was not set up with the
correct and/or consistent reference data such as statevector, 
time correlation and/or GHA correction.  This problem may be confirmed by 
examining RDS debug log.  If the log file showed errors like
"Pulse YYYY-DDDTHH:MM:SS.UUUUUU too far away"
or "get_dops() failed in {near|mid|far} range" for this job, then 
definitely, at least one of the aforementioned data elements is
incorrect or at the least inconsistent with the data found.
(3) And last but not unlikely reason is there is just not enough data for
\fBrds\fP to produce one of the fixed-center frames.  \fBrds\fP does not
generate partial image frame.  Recommended action is 
to double-check the tape mounted and the area scanned.  
Ascertain that the setup or configuration parameters agree with those 
reported for this downlink.
.IP "\fBTape label mismatch\fP"
The tape label specified in the job request does not match with what 
recorded about the scanned area on the mounted tape.  Recomended action is to
make sure that (1) the correct tape was mounted, (2) the tape area scanned
does have some recorded label that matched with what specified in 
the job request.  Recommended action here is to look in RDS debug log for the
response string \fIDU R-\fR or \fIDU F-\fR that followed the \fIPL\fR command.
What immediately follows the 5 characters "\fIDU R-\fR" (or "\fIDU F-\fR") and
comes before the ";" is the recorded user data.  Check to see if this
recorded user data contains a substring that exactly matched the label text
specified in the job request.  If not, modify the job parameter.
.IP "\fBNo tape in drive\fP"
If for any reason \fBrds\fP failed to receive a valid response for the \fIPL\fR
command, it will issue a \fIDF\fR command to see whether the tape
tape was mounted properly.  If there is no response or the response indicates
that the current tape footage is anything other than non-zero,
\fBrds\fP will automatically assume that the tape is not mounted.  However, if
visual inspection showed otherwise,
check the Error Correction (ECC) Unit.  Make sure it is powered on. 
.IP "\fBNo memory for RDS_t object\fP"
This error message is generated only if \fBrds\fP failed to initialize itself
during start-up.
(1) Often, this is indicative of the existence of either some zombie or 
unterminated memory-hogged processes in the system such as \fBrds\fP at the 
time of startup.  Recommended action is to determine what that process
is and manually kill it.  Then restart \fBrds\fP.  (2) Otherwise, the problem
may be caused by improper setting of some RDS configuration parameters, which 
can result in excessive memory allocation.  Double-check all the parameters
in RDS configuration file and restart \fBrds\fP.
.IP "\fBToo many tape blocks\fP"
The total number of tape blocks requested to process exceeds the limit 
set by the RDS configurable parameter \fBTAPEBLK_MAX\fP.  To proceed with
this job, modify the parameter accordingly using an editor,
reset \fBrds\fP through CP, and then resubmit the job.
.IP "\fBInvalid or missing BODY.START_BLOCK\fP"
Check the datatake's starting tape address.  It must be a positive integer,
i.e., strictly greater than 0.  If not, modify the address and then resubmit
the job.
.IP "\fBInvalid or missing BODY.END_BLOCK\fP"
Check the datatake's end tape address.  It must be a positive integer
greater than or equal to the starting tape address.  If not, modify the 
address.  Then, resubmit the job.
.IP "\fBCan't create <filename>: <error reason>\fP"
\fBrds\fP was unable to create its output file because of some error.
The cause of the error should be self-explanatory from the message text.
The common cause is lack of write permission or privilege.
.IP "\fBCan't write to <filename>: <error reason>\fP"
\fBrds\fP was unable to write to its output file for some reasons.
The cause of the error should be self-explanatory from the message text.
Common causes are the disk filesystem full and permission denied.
.IP "\fBCan't rename <filename>: <error reason>\fP"
This error occurs only during the \fIclean-up\fR phase of error proccesing,
where \fBrds\fP was unable to save its intermediate output files to the
specified directory for post-mortem analysis.  Check the write permission
of the directory and make sure that it is writeable by \fBrds\fP.
Then, manually move the file yourself.
.IP "\fBCan't delete <filename>: <error reason>\fP"
This error occurs only during the \fIclean-up\fR phase of error processing,
where \fBrds\fP was unable to remove its intermediate output file after it
has successfully copied to the specified directory area.
Recommended action here is to manually delete the intermediate file yourself.
.IP "\fBInvalid or no response to 'RS'\fP"
There are two possible causes for this error.  First, (1) the most likely one
is the DCRSi drive was not powered on or its connection to the host was
broken.  To confirm, look in RDS debug log for the DCRSi response to the 
\fIRS\fR command.  If there is no response, then it is very likely that this is
the cause. (2) Otherwise, if the response appears to be an undecipherable, 
random stream of characters, then the serial control line to the DCRSi drive
was not initialized properly by \fBrds\fP due to the mismatch between 
the DCRSi switch setting and the configuration parameter \fBDCRSI_CMD_CSIZE\fP.
Make sure that the two match.
.IP "\fBSTATE_VECTOR_DATA.TIME must be <= BODY.START_TIME\fP"
The reference statevector time specified in the job request must be the
same or before the datatake's start time. Though theoretically possible,
\fBrds\fP does not backward propagate the initial reference statevector.
The recommendation here is change the datatake's start time.
.SH DCRSi COMMANDS & RESPONSES
This section lists all of the DCRSi command sequences sent to and 
response patterns expected back from the DCRSi by \fBrds\fP.
They are listed in order of usage.  Each command sequence is actually
transmitted with enclosing semicolons.  The response string
received must contain all of the expected patterns and in the correct order 
before the response is considered valid and complete and, if any, before the 
next command sequence is sent.
If there is more than one expected, the response patterns are shown
separated by commas.  If only one of the patterns expected, they are shown
separated by vertical bars.  Braces are used to delineate the descriptive
text only, not part of transmission.
The exact commands transmitted are shown highlighted in the left column, 
whereas the responses expected are shown underlined.
.TP 15
.B RS
{\fIRevision\fR, {{\fIDS 4000\fR} | {\fIDS 4011\fR}}}
.TP
.B EE;PD
\fI**\fR
.TP
.B PL A1234567
{\fIDL B-1234567\fR, {\fIDS C022\fR | {\fIDS 4011\fR}}
.TP
.B DU
{{\fIDU R-\fR{label}\fI;*\fR} | {\fIDU F-\fR{label}\fI;*\fR}}
.TP
.B DF
{\fIDF 1234\fR}
.TP
.B SL
{{\fIDS 4011\fR} | {\fIDS 4000\fR}}
.TP
.B MD;UL
\fI*\fR
.TP 0
where \fB1234567\fP and \fI1234\fR denote the starting address to 
playback and the current tape footage, respectively.
.SH DCRSi VERIFICATION 
To ensure data integrity and output consistency from the DCRSi drive,
the following verification procedure should be performed at some 
self-established confidence interval.  We recommend every other 
day to start.  The procedure is simple and should take only few minutes
to complete.  Have RDS perform at least one scan of a known good tape area
and compare its new output against an existing, already established,
good scan results file.  If there is any difference, then it is probably about
time to recalibrate the DCRSi drive and/or the Error Correction Unit.
.\" Induce a page break here.
.SH ""
.SH ""
.SH ""
.SH ENVIRONMENT
The environment variable ASF should be set to \fI/ASF/XXX\fR,
where \fIXXX\fR is the name
of the directory corresponding to the software release being used, e.g.,
\fIR1BP\fR for the R1B' delivery.

Similarly, the environment variable LOCAL should be set to \fI/LOCAL/XXX\fR,
where \fIXXX\fR equals \fIR1BP\fR for the R1B' delivery.

If either ASF or LOCAL is not set, \fBrds\fP prints an error message and exits.

.SH FILES
.PD 0
.TP 27
.B $LOCAL/rds/rds-config.odl
File containing configurable parameters for \fBrds\fP.  Please see the
\fBCONFIGURATION PARAMETERS\fP section for details on the parameters.
.TP
.B $LOCAL/sps/templates/msgTemplate.odl
Name of the message template file that defines the formats of all ODL
messages recognized by \fBrds\fP
(i.e., \fI$LOCAL/sps/templates/msgTemplate.odl\fP).
.TP
.B /dev/vdi
Data interface to the DCRSi drive
.TP
.B /dev/vdi_sp0
(Serial) control interface to the DCRSi drive
.SH "SEE ALSO"
CP(1), CP_scan_qc(1), log_browser(1), taperead(1)

.\".SH NOTES
.SH REFERENCES
\fIRaw Data Scanner Software Operator's Manual\fR, JPL D-13577, Version 1.0,
May 1996.

\fIPlanetary Data System Standards Reference\fR, Part 2,
Version 3.0, JPL D-7669, November 1992.  This document contains
information about ODL (Object Description Language) syntax; both
the RDS configuration file and message template are written in ODL.

.\".SH CAVEATS
.\".SH DIAGNOSTICS
.\"all the possible error messages the program can print out, and
.\"what they mean.
.\".SH BUGS
.\".SH RESTRICTIONS
.\"bugs you don't plan to fix :-)
.SH AUTHOR
T. Truong
.\" .SH HISTORY
.\" programs derived from other sources sometimes have this.

.SH CREDITS
This software was developed as part of the Raw Data Scanner Subsystem
of the Alaska SAR Facility at the University of Alaska, Fairbanks.


unsupported/asf_ops_R2.1.2/rds/src/rds/rds.c
/*=============================================================================
 |  @(#)rds.c	2.51 98/03/15 12:04:47
 |
 |  Multithreaded Raw Data Scanner (RDS) Main Program.
 |  Copyright (C) Jet Propulsion Laboratory.
 |  Alaska SAR Facility (ASF) Project.
 |
 *============================================================================*/
#include <sys/types.h>
#include <syslog.h>
#include <time.h>
#include "RDS.h"
#include "logUtils.h"
#include "asf_syslog.h"
#include "asf.h"
#include "version.h"

static const char sccsid_rds_c[] =
        "@(#)rds.c	2.51 98/03/15 12:04:47";

static const char copyright[] = "\
Copyright (C) 1996, California Institute of Technology.  \
ALL RIGHTS RESERVED.  U.S. Government Sponsorship acknowledged.";

static int processing = 0;
static int resetting = 0;
static int off_line = 0;

/*============================================================================*
 |  NAME
 |
 |  SYPNOSIS
 |
 |  DESCRIPTION
 |
 *============================================================================*/
static
int SendAckToServer (ODL msg, char* ack_type, char* status, char* text)
{
    char s[80];
    ODL ack;
    register int n, c = 0;
    if (off_line) return 0;

    if (! (ack = GetAckMsg(msg, ack_type))) {
        printfLLog(LOG_ERR, "SendAckToServer() can't retrieve %s", ack_type);
        return -1;
    }
    n = strlen(text)-1;
    if (text[n] == '\n') {
	c = text[n]; text[n] = 0;
    }
    ODLSetVal(ack, (sprintf(s,"%s.BODY.STATUS", ack_type),s), status);
    ODLSetVal(ack, (sprintf(s,"%s.BODY.COMMENT",ack_type),s), text);

    if (c) text[n] = c;

    if (!strcmp(ack_type, "SUBSYSTEM_STATUS"))
        ODLSetVal(ack, (sprintf(s,"%s.BODY.SUB_TYPE",ack_type),s), "HEARTBEAT");

    if (WriteMsgToServer(ack) == -1) {
	printfLLog(LOG_ERR, "Unable to send %s to CP", ack_type);
	ODLFree(ack);
	return -1;
    }
    printfLLog(LOG_DEBUG, "%s sent to CP", ack_type);
    ODLFree(ack);
    return 0;
}

/*============================================================================*
 |  NAME
 |
 |  SYPNOSIS
 |
 |  DESCRIPTION
 |
 *============================================================================*/
static
void heartbeat_handler(ODL msg, RDS_t* rds)
{
    int  blk;
    char buf[80];
    
    if (resetting) 
        sprintf(buf, "Resetting job %d...", rds->job.job_id);

    else if (! processing ||
       (blk = RDS_Position(rds) - rds->job.start.blk,
	blk >= rds->job.end.blk - rds->job.start.blk))

	sprintf(buf, rds->pulse_to_check
	    ? "Looking for PCM SYNC to determine segment time..."
	    : "Waiting for a processing request...", rds->pulse_to_check);
    else if (blk == 0)

        sprintf (buf, "Seeking block #%d", rds->job.start.blk);
    else
        sprintf(buf, "%d%% completed - %s %d..%d", (int)
                ((blk * 100.0) / (rds->job.end.blk - rds->job.start.blk)),
		rds->job.scan_job ? "Scanning" : "Decoding",
                BLKNUM(rds->job.start.blk+blk, &rds->job, &rds->job),
		BLKNUM(rds->job.end.blk, &rds->job, &rds->job));
    SendAckToServer(msg, "SUBSYSTEM_STATUS", "COMPLETED", buf);
}

/*============================================================================*
 |  NAME
 |
 |  SYPNOSIS
 |
 |  DESCRIPTION
 |
 *============================================================================*/
static
void job_handler(ODL msg, RDS_t* rds)
{
    const Job_t* job = &rds->job;

    SendAckToServer(msg, "SUBSYSTEM_ACK", "COMPLETED", "Greetings from RDS!");
    processing = 1;

    if (RDS_Init(rds, Value(msg))) {
	RDS_Scan(rds);
	RDS_Destroy(rds);
    }
    processing = 0;

    if (! resetting) {
	char  err[256];
        const char* cp = RDS_Error(rds);
	const char* np = strchr(cp, '\n');
	const int n = (!np || !(np = strchr(np+1,'\n'))) ? strlen(cp) : (np-cp);

	memcpy(err, cp, n>sizeof(err)-1 ? sizeof(err)-1 : n);
	err[n] = 0;
	
	if (err[0])
	    printfLLog(LOG_ERR, "Job %d done - %s", rds->job.job_id, err);
	else
	    printfLLog(LOG_ERR, "Job %d done - OK", rds->job.job_id);

	SendAckToServer(msg, "SUBSYSTEM_COMPLETED", 
	    (strstr(cp, "tape label") ? "WRONG_TAPE" :
	    (RDS_Errno(rds) == -1 ? "CANCEL/FAIL" : "COMPLETED")), err);
    }
}

/*============================================================================*
 |  NAME
 |
 |  SYPNOSIS
 |
 |  DESCRIPTION
 |
 *============================================================================*/
static
void cleanup_handler(ODL msg, RDS_t* rds)
{
    char s[64], err[256], *p, *ack_type = "SUBSYSTEM_STATUS";
    register ODL ack = Value(msg);
    register char* status =
	RDS_Cleanup( rds,
	    ((p = ODLGetStr(ack, "BODY.SAVE_FLAG")) && !strcasecmp(p, "YES") ?
	    ODLGetStr(ack, "BODY.SAVE_DIR") : NULL), err)
	== -1 ? "CANCEL/FAIL" : "COMPLETED";

    if (! (ack = GetAckMsg(msg, ack_type))) {
        printfLLog(LOG_ERR, "cleanup_handler: can't retrieve %s", ack_type);
        return;
    }
    ODLSetVal(ack, (sprintf(s,"%s.BODY.STATUS",ack_type), s), status);
    ODLSetVal(ack, (sprintf(s,"%s.BODY.COMMENT",ack_type), s), err);
    ODLSetVal(ack, (sprintf(s,"%s.BODY.SUB_TYPE",ack_type),s), "CLEANUP");

    WriteMsgToServer(ack);
    ODLFree(ack);
}

/*============================================================================*
 |  NAME
 |
 |  SYPNOSIS
 |
 |  DESCRIPTION
 |
 *============================================================================*/
static
void reset_handler(ODL msg, RDS_t* rds)
{
    resetting = 1;
    if (processing) {
	printfLLog(LOG_ERR, "Cancelling job %d... ", rds->job.job_id);
	RDS_Reset(rds);
	printfLLog(LOG_ERR, "Job %d cancelled", rds->job.job_id);
    }
    resetting = 0;
}

/*============================================================================*
 |  NAME
 |
 |  SYPNOSIS
 |
 |  DESCRIPTION
 |
 *============================================================================*/
int main(int argc, char *argv[])
{
    RDS_t* rds;
    time_t clock;
    char err[256], ts[28];
    register int i = 0;

    ctime_r((time(&clock), &clock), ts, sizeof(ts));
    ts[strlen(ts)-1] = 0;
    openlog(argv[0], LOG_PID|LOG_CONS|LOG_NDELAY, LOG_SPS);

    while (++i < argc && strcmp(argv[i], "-configfile"));
    if (++i >= argc) {
	syslog(LOG_ERR, "%s: configfile parameter missing", version_id);
	printf("%s %s: configfile parameter missing\n", ts, version_id);
	exit(-1);
    }
    if (! (rds = RDS_malloc(argv[i], err))) {
        syslog(LOG_ERR, "%s: %s",  version_id, err);
        printf("%s %s: %s\n",  ts, version_id, err);
        exit(-1);
    }
    for (i = 1; i < argc && strcmp(argv[i], "-jobfile"); ++i);
    if (i == argc) {
	/*  We are running with CP */

	AsfApp app;
	if (! (app = AsfAppInitialize("RDS", &argc, argv))) {
	    syslog(LOG_ERR, "%s: AsfAppInitialize() failed.",  version_id);
	    printf("%s %s: AsfAppInitialize() failed.\n",  ts, version_id);
	    exit (-1);
	}
	printfLLog(LOG_INFO, "%s - Started %s...", version_id, ts);
	printfLLog(LOG_INFO, "%s", copyright);

	AsfAddCallback(app, "SPS_SCAN_REQUEST", job_handler, rds,
	    "SPS_DECODE_REQUEST", job_handler, rds,
	    "SUBSYSTEM_HEARTBEAT", heartbeat_handler, rds,
            "SUBSYSTEM_HALT", reset_handler, rds,
            "SUBSYSTEM_STOP", reset_handler, rds,
            "CLEANUP_REQUEST", cleanup_handler, rds,
	    NULL);
	AsfAppMainLoop(app);
    }
    else if (++i >= argc) {
	syslog(LOG_ERR, "%s: jobfile parameter missing", version_id);
	printf("%s %s: jobfile parameter missing\n", ts, version_id);
	exit(-1);
    }
    else { /*  We are running without CP */

	register int k;
	ODL odl;

	if (initASFlog(argc, argv, "RDS") == -1) {
	    syslog(LOG_ERR, "%s: initASFlog() failed.",  version_id);
	    printf("%s %s: AsfAppInitialize() failed.\n",  ts, version_id);
	    exit (-1);
	}
	for (k = 1; k < argc && strcmp(argv[k], "-config"); ++k);
	if (k == argc) {
	    printfLLog(LOG_ERR, "%s: -config parameter missing", version_id);
	    exit(-1);
	}
	if (! InitODLtemplate(argv[++k])) {
	    printfLLog(LOG_ERR, "%s: InitODLtemplate failed.", version_id);
	    exit (-1);
	}
	printfLLog(LOG_INFO, "%s - Started %s...", version_id, ts);
	printfLLog(LOG_INFO, "%s", copyright);
	off_line = 1;

	while (i < argc && *argv[i] != '-') {
	    if (! (odl = ODLparse(argv[i], 0, err))) {
		printfLLog(LOG_ERR, "%s: %s, %s", version_id, argv[i], err);
		exit (-1);
	    }
	    job_handler(odl, rds);
	    ODLFree(odl);
	    i++;
	}
    }
    RDS_free(rds);
}
